### ISSUE TITLE
composeLoggerConfig crashes with lodash <= 4.17.10

### ISSUE THREAD

**Issue by tyronet-sportsbet** [wix/Detox#3849]
### What happened?

Detox does not function at all if the package manager resolves `lodash` to 4.17.10 (and possibly earlier). It raises a syntax error when starting up the DetoxSecondaryContext.

### What was the expected behaviour?

Not to crash :smile:

### Was it tested on latest Detox?

- [X] I have tested this issue on the latest Detox release and it still reproduces.

### Help us reproduce this issue!

Please see here for a minimal reproduction repo: https://github.com/tyronet-sportsbet/DetoxLodashIssue.

I tried to use your DetoxTemplate project, however:
- It's still using Detox 19.x
- It's still using React Native 0.66.x
- The `yarn.lock` references `https://repo.dev.wixpress.com` all over the place, which is inaccessible from my network, making the template unusable 😢.

My template is just a brand new react-native template project built from 0.71.0 with `jest@^29` and `detox@^20` added. I've used a yarn resolution to _force_ it to use `lodash@4.17.10` in order to reproduce the issue.

### In what environment did this happen?

Detox version: 20.1.1
React Native version: 0.71.0
Has Fabric (React Native's new rendering system) enabled: no
Node version: 19.4.0 (also reproduced on 18.13.0 and 16.3.0)
Test-runner (select one): jest

### Detox logs

<details><summary>Detox logs</summary>

```
yarn run v1.22.18
$ /Users/tyronet/Dev/DetoxLodashIssue/node_modules/.bin/detox test --loglevel trace -c ios.sim.debug e2e/starter.test.js
15:44:26.587 detox[59709] B lifecycle node_modules/.bin/detox test --loglevel trace -c ios.sim.debug e2e/starter.test.js
  data: {
    "id": "170d792a-0dbd-104c-5013-9a3993812ed2",
    "detoxConfig": {
      "configurationName": "ios.sim.debug",
      "apps": {
        "default": {
          "type": "ios.app",
          "binaryPath": "ios/build/Build/Products/Debug-iphonesimulator/DetoxLodashIssue.app",
          "build": "xcodebuild -workspace ios/DetoxLodashIssue.xcworkspace -scheme DetoxLodashIssue -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build"
        }
      },
      "artifacts": {
        "rootDir": "artifacts/ios.sim.debug.2023-01-18 04-44-26Z",
        "plugins": {
          "log": {
            "enabled": false,
            "keepOnlyFailedTestsArtifacts": false
          },
          "screenshot": {
            "enabled": true,
            "shouldTakeAutomaticSnapshots": false,
            "keepOnlyFailedTestsArtifacts": false
          },
          "video": {
            "enabled": false,
            "keepOnlyFailedTestsArtifacts": false
          },
          "instruments": {
            "enabled": false,
            "keepOnlyFailedTestsArtifacts": false
          },
          "uiHierarchy": {
            "enabled": false,
            "keepOnlyFailedTestsArtifacts": false
          }
        }
      },
      "behavior": {
        "init": {
          "keepLockFile": false,
          "reinstallApp": true,
          "exposeGlobals": true
        },
        "cleanup": {
          "shutdownDevice": false
        },
        "launchApp": "auto"
      },
      "cli": {
        "configuration": "ios.sim.debug",
        "loglevel": "trace"
      },
      "device": {
        "type": "ios.simulator",
        "device": {
          "type": "iPhone 12"
        }
      },
      "logger": {
        "level": "trace",
        "overrideConsole": true
      },
      "testRunner": {
        "retries": 0,
        "forwardEnv": false,
        "bail": false,
        "jest": {
          "setupTimeout": 120000,
          "teardownTimeout": 30000,
          "retryAfterCircusRetries": false,
          "reportWorkerAssign": true
        },
        "args": {
          "$0": "jest",
          "_": [
            "e2e/starter.test.js"
          ],
          "config": "e2e/jest.config.js",
          "--": []
        }
      },
      "session": {
        "autoStart": true,
        "debugSynchronization": 10000
      }
    },
    "detoxIPCServer": "primary-59709",
    "testResults": [],
    "testSessionIndex": 0,
    "workersCount": 0
  }
15:44:26.596 detox[59709] i ipc Server path not specified, so defaulting to ipc.config.socketRoot + ipc.config.appspace + ipc.config.id /tmp/detox.primary-59709
15:44:26.597 detox[59709] i ipc starting server on  /tmp/detox.primary-59709 
15:44:26.597 detox[59709] i ipc starting TLS server false
15:44:26.597 detox[59709] i ipc starting server as Unix || Windows Socket
15:44:26.605 detox[59709] i ws-server Detox server listening on localhost:49371...
15:44:26.607 detox[59709] i lifecycle Serialized the session state at: /private/var/folders/y5/xf1p97wn1r5g9chr3qqxgjf80000gn/T/170d792a-0dbd-104c-5013-9a3993812ed2.detox.json
15:44:26.608 detox[59709] B lifecycle jest --config e2e/jest.config.js e2e/starter.test.js
evalmachine.<anonymous>:1
(defaultOptions({ level }) {
                           ^

SyntaxError: Unexpected token '{'
    at new Script (node:vm:100:7)
    at createScript (node:vm:258:10)
    at Object.runInContext (node:vm:289:10)
    at Object._reviver (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src/ipc/SessionState.js:45:17)
    at JSON.parse (<anonymous>)
    at Object.parseJC [as parse] (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/json-cycle/cycle.js:136:28)
    at SessionState.parse (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src/ipc/SessionState.js:40:28)
    at [restoreSessionState] (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src/realms/DetoxSecondaryContext.js:86:25)
    at new DetoxContext (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src/realms/DetoxContext.js:47:53)
    at new DetoxSecondaryContext (/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src/realms/DetoxSecondaryContext.js:18:5)
15:44:27.324 detox[59709] E lifecycle Command failed with exit code = 1:
jest --config e2e/jest.config.js e2e/starter.test.js
15:44:27.324 detox[59709] i ws-server Detox server has been closed gracefully
15:44:27.326 detox[59709] E lifecycle node_modules/.bin/detox test --loglevel trace -c ios.sim.debug e2e/starter.test.js
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

</details>

### Device logs

Omitted. Detox doesn't get far enough to run anything on the device or simulator.

### More data, please!

If I edit `composeLoggerConfig.js` to add `console.log` statements to the `composeLoggerConfig` function, like this:

```diff
@@ -29,6 +29,7 @@ function composeLoggerConfig(opts) {
      * @param config
      */
     (acc, config) => {
+      console.log("acc = ", acc, "config = ", config)
       if (!config) return acc;
       const { options } = config;
       return _.merge(acc, {
```

Then on lodash 4.17.10 (broken) it prints out this last:

```
acc =  {
  level: 'trace',
  overrideConsole: true,
  options: [Function: defaultOptions] {
    showDate: [Function: shortFormat],
    showLoggerName: true,
    showPid: true,
    showLevel: false,
    showMetadata: false,
    showPrefixes: [Function: showPrefixes],
    basepath: '/Users/tyronet/Dev/DetoxLodashIssue/node_modules/detox/src',
    prefixers: {
      ph: [Function (anonymous)],
      cat: [Function (anonymous)],
      event: [Function (anonymous)],
      id: [Function (anonymous)]
    },
    stringifiers: {
      args: [Function: args],
      error: [Function: identity],
      data: [Function: data],
      stack: [Function: identity],
      origin: [Function: identity]
    }
  }
} config =  { level: 'trace' }
```

Note how `options` is still a **function** here.

In 4.17.11 (fixed) I see:

```
acc =  {
  level: 'trace',
  overrideConsole: true,
  options: {
    showDate: [Function: shortFormat],
    showLoggerName: true,
    showPid: true,
    showLevel: false,
    showMetadata: false,
    showPrefixes: [Function: showPrefixes],
    basepath: '/Users/tyronet/Dev/hydra/node_modules/detox/src',
    prefixers: {
      ph: [Function (anonymous)],
      cat: [Function (anonymous)],
      event: [Function (anonymous)],
      id: [Function (anonymous)]
    },
    stringifiers: {
      args: [Function: args],
      error: [Function: identity],
      data: [Function: data],
      stack: [Function: identity],
      origin: [Function: identity]
    }
  }
} config =  { level: 'trace' }
```

Note how it's an **object** now, not a function.

## Easy Resolution

Detox currently marks its lodash requirement as `lodash@^4.17.5`, however as demonstrated by this issue, Detox doesn't behave unless lodash is _at least_ `4.17.11`. I suggest changing in `detox/package.json` your dependency on lodash to `lodash@^4.17.11` to prevent package managers from resolving Detox's lodash to a version that doesn't actually work with it.

I'm happy to raise a PR with this one-line change if you'd like, I'm just raising the issue first as per the contributing guidelines.

**Comment by noomorph** [comment]
Thanks, although I'm already curious why it crashes. I'll try to reproduce but I'd be happy if you assist me in case if complications arise.

I get your idea about a one-liner fix, but let's do our best to find out what in fact was wrong.

**Comment by noomorph** [comment]
> I tried to use your DetoxTemplate project, however:

> It's still using Detox 19.x
> It's still using React Native 0.66.x
> The yarn.lock references https://repo.dev.wixpress.com all over the place, which is inaccessible from my network, making the template unusable 😢.
> My template is just a brand new react-native template project built from 0.71.0 with jest@^29 and detox@^20 added. I've used a yarn resolution to force it to use lodash@4.17.10 in order to reproduce the issue.

Thanks for the report! Could you file it for us as a separate issue here? 🙏

**Comment by tyronet-sportsbet** [comment]
> Thanks, although I'm already curious why it crashes. I'll try to reproduce but I'd be happy if you assist me in case if complications arise.

Happy to help any way I can. It should be very easy to reproduce, simply clone [my mimimal reproduction repo](https://github.com/tyronet-sportsbet/DetoxLodashIssue), make sure to use `yarn` to `yarn install`. Building is as per any barebones react native project, e.g. iOS:

```
cd ios
pod install
cd ..
yarn detox build -c ios.sim.debug
```
Then you'll see the crash on: `yarn detox test --loglevel trace -c ios.sim.debug e2e/starter.test.js`.

> Thanks for the report! Could you file it for us as a separate issue here? 🙏

Do you mean file an issue on https://github.com/wix-incubator/DetoxTemplate? If so, will do.

**Comment by noomorph** [comment]
@tyronet-sportsbet please see #3851 and my comment there below about creating an extra PR.

**Comment by noomorph** [comment]
Resolved in Detox 20.1.2
